### DISCUSSION TITLE
Platform-specific addons

### DISCUSSION THREAD

**Discussion by iqnite** [ScratchAddons/ScratchAddons#7392] [Ideas]
How would we handle addons and features that only work in certain browsers/platforms? I can think of the following scenarios:
- IIRC the "Wrap list items" addon was canceled because of a rendering bug in Google Chrome. It might work in other browsers.
- Apple is opening its App Store, so Safari is a growing platform Scratch Addons may be ported to. Developing the same extension for 3 different browsers can be tricky, especially for backend functionalities, and compromises can't always be found.
- With the growing amount of tablet users, we might also think of a mobile version of Scratch Addons. Tablets handle fullscreen differently than PCs, so the "Enhanced fullscreen" addon might lead to unexpected behaviors.
- ~~TurboWarp uses a selection of compatible addons. This requires manual copy-paste work with every update, plus design inconsistencies between Scratch Addons' and TurboWarp's settings page.~~

How this would be implemented:
- All addons receive hidden tags for compatible target platforms (Browser [Chromium, Firefox, ...], Platform [Desktop, Tablet, ...], TurboWarp [true/false], ...)
- To be shown in the settings page, an addon must be compatible with the current platform
- If it's not, it's shown in a separate category at the bottom of the settings page, with a message telling the user that the addon is incompatible with their device.
- Maybe there could be an advanced setting allowing users to bypass compatibility checks and enable addons on unsupported platforms (at their own risk)?

**Comment by CST1229** [comment]
> The "Wrap list items" addon was canceled because of a rendering bug in Google Chrome. In Firefox, it would have worked.

I thought it was due to how Scratch only renders the visible areas of a list, with the calculations assuming a fixed row height?

**Reply by WorldLanguages** [reply]
I thought it was broken on Firefox too.

**Comment by WorldLanguages** [comment]
The idea sounds interesting, it would be specially useful to distinguish addons that support touchscreen usage vs those that don't.

**Reply by Secret-chest** [reply]
What if you use a mouse on phone or tablet?

**Reply by Samq64** [reply]
A toggle in more settings to always show all addons should solve that.

**Reply by DNin01** [reply]
That's fine as long as we still try to, wherever possible, add support for touch control.

> What if you use a mouse on phone or tablet?

AFAIK there are ways you can detect these peripherals, and some may not even require a user gesture.

**Reply by BroJac5246** [reply]
> That's fine as long as we still try to, wherever possible, add support for touch control.
> 
> > What if you use a mouse on phone or tablet?
> 
> AFAIK there are ways you can detect these peripherals, and some may not even require a user gesture.

We could also display a banner saying that we're hiding them. That way people will be aware that there are hidden addons, even if they set up the extension using a touchscreen.

**Comment by DNin01** [comment]
> IIRC the "Wrap list items" addon was canceled because of a rendering bug in Google Chrome. It might work in other browsers.

If we can't implement a feature in the Chrome version of our extension, then it probably shouldn't be implemented for any other browser either.

In general, I think we should prefer releasing features to all browsers and platforms at the same time. But if a suite of features proves to be incompatible with a different type of device or operating system, that's when it makes more sense to leave something out.

**Comment by GarboMuffin** [comment]
> TurboWarp uses a selection of compatible addons. This requires manual copy-paste work with every update, plus design inconsistencies between Scratch Addons' and TurboWarp's settings page.

The only thing being copy-pasted is addon IDs. Scripts generate the rest of the addon files. Having this metadata live in TurboWarp is much more convenient for everyone.

The settings pages have very different technical requirements (eg. vue vs react, browser extension vs web, webpack, etc.) and the designs don't need to match (in some cases deviation is quite intentional). I don't see a reason to merge them.
